sbi/inference/npse/npse.py

```diff
@@ -0,0 +1,612 @@
+                self._summary["training_loss"].append(train_loss_average)
+            else:
+                previous_loss = self._summary["training_loss"][-1]
+                self._summary["training_loss"].append(
+                    (1.0 - ema_loss_decay) * previous_loss
+                    + ema_loss_decay * train_loss_average
+                )
+
+            # Calculate validation performance.
+            self._neural_net.eval()
+            val_loss_sum = 0
+
+            with torch.no_grad():
+                for batch in val_loader:
+                    theta_batch, x_batch, masks_batch = (
+                        batch[0].to(self._device),
+                        batch[1].to(self._device),
+                        batch[2].to(self._device),
+                    )
+                    # Take negative loss here to get validation log_prob.
+                    val_losses = self._loss(
+                        theta_batch,
+                        x_batch,
+                        masks_batch,
+                        proposal,
+                        calibration_kernel,
+                        force_first_round_loss=force_first_round_loss,
+                    )
+                    val_loss_sum += val_losses.sum().item()
+
+            # Take mean over all validation samples.
+            val_loss = val_loss_sum / (
+                len(val_loader) * val_loader.batch_size  # type: ignore
+            )
+
+            # NOTE: Due to the inherently noisy nature we do instead log a exponential
+            # moving average of the validation loss.
+            if len(self._summary["validation_loss"]) == 0:
+                val_loss_ema = val_loss
+            else:
+                previous_loss = self._summary["validation_loss"][-1]
+                val_loss_ema = (
+                    1 - ema_loss_decay
+                ) * previous_loss + ema_loss_decay * val_loss
+
+            self._val_loss = val_loss_ema
+            self._summary["validation_loss"].append(self._val_loss)
+            self._summary["epoch_durations_sec"].append(time.time() - epoch_start_time)
+
+            self._maybe_show_progress(self._show_progress_bars, self.epoch)
+
+        self._report_convergence_at_end(self.epoch, stop_after_epochs, max_num_epochs)
+
+        # Update summary.
+        self._summary["epochs_trained"].append(self.epoch)
+        self._summary["best_validation_loss"].append(self._val_loss)
+
+        # Update tensorboard and summary dict.
+        self._summarize(round_=self._round)
+
+        # Update description for progress bar.
+        if show_train_summary:
+            print(self._describe_round(self._round, self._summary))
+
+        # Avoid keeping the gradients in the resulting network, which can
+        # cause memory leakage when benchmarking.
+        self._neural_net.zero_grad(set_to_none=True)
+
+        return deepcopy(self._neural_net)
+
+    def _converged(self, epoch: int, stop_after_epochs: int) -> bool:
```

Comment:
This is odd to me. Almost the same method exists in `base.py`, but with the difference that it writes to `_best_model_state_dict`, which this method does not seem to do? So are we not resetting to the best model?